File: 21MergeTwoSortedLists.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	var output *ListNode
	curNodeL1 := l1
	curNodeL2 := l2
	curNodeOutput := output

	if curNodeL1 != nil || curNodeL2 != nil {
		output = &ListNode{}
		curNodeOutput = output
	}

	for {
		if curNodeL1 != nil && curNodeL2 != nil {
			if curNodeL1.Val <= curNodeL2.Val {
				curNodeOutput.Val = curNodeL1.Val
				curNodeL1 = curNodeL1.Next
			} else {
				curNodeOutput.Val = curNodeL2.Val
				curNodeL2 = curNodeL2.Next
			}
		} else if curNodeL1 == nil && curNodeL2 != nil {
			curNodeOutput.Val = curNodeL2.Val
			curNodeL2 = curNodeL2.Next
		} else if curNodeL1 != nil && curNodeL2 == nil {
			curNodeOutput.Val = curNodeL1.Val
			curNodeL1 = curNodeL1.Next
		} else {
			break
		}

		if curNodeL1 != nil || curNodeL2 != nil {
			curNodeOutput.Next = &ListNode{}
			curNodeOutput = curNodeOutput.Next
		}
	}

	return output
}

func main() {
	output := mergeTwoLists(&ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 4, Next: nil}}},
		&ListNode{Val: 1, Next: &ListNode{Val: 3, Next: &ListNode{Val: 4, Next: nil}}})
	output1 := mergeTwoLists(&ListNode{}, &ListNode{})
	output2 := mergeTwoLists(&ListNode{}, &ListNode{Val: 0, Next: nil})

	fmt.Println("\n******** output *********\n", output, output1, output2)
}

File: 21MergeTwoSortedLists.cpp
#include <iostream>
#include <set>
#include <vector>
#include <iterator>
#include <utility>
#include <cstdlib>
#include <limits>
#include <functional>
#include <stack>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <algorithm>
#include <cstdint>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

template<class T>
void print2DimVec(const vector<vector<T>>& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[" << endl;
        }
        printContainer(*it);
        if (it == prev(container.end())) {
            cout << "]" << endl;
        }
    }
}

template<class T>
void printMap(const T& container) {
//    return;

    cout << "size " << container.size() << endl;
    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }
    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[ ";
        }

        cout << "[" << it->first << ", " << it->second << "]";

        if (next(it) == container.end()) {
            cout << " ]" << endl;
        } else {
            cout << ", ";
        }
    }
}


/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

void printList(const ListNode* node) {
    while (node != nullptr) {
        cout << node->val;

        node = node->next;
        
        if (node != nullptr) {
            cout << "->";
        } else {
            cout << endl;
        }
    }
}

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (l1 == nullptr && l2 == nullptr) {
            return nullptr;
        }
        ListNode* root = new ListNode;
        ListNode* cur = root;

        while (l1 != nullptr || l2 != nullptr) {
            if (l1 != nullptr && l2 != nullptr) {
                if (l1->val <= l2->val) {
                    cur->val = l1->val;
                    l1 = l1->next;
                } else {
                    cur->val = l2->val;
                    l2 = l2->next;
                }
            } else if (l1 == nullptr && l2 != nullptr) {
                cur->val = l2->val;
                l2 = l2->next;
            } else if (l1 != nullptr && l2 == nullptr) {
                cur->val = l1->val;
                l1 = l1->next;
            } else if (l1 == nullptr && l2 == nullptr) {
                break;
            }

            if (l1 != nullptr || l2 != nullptr) {
                cur->next = new ListNode;
                cur = cur->next;
            }
        }

        return root;
    }
};

int main(){
    Solution s;

	auto output = s.mergeTwoLists(new ListNode(1, new ListNode(2, new ListNode(4))),
            new ListNode(1, new ListNode(3, new ListNode(4))));
	auto output1 = s.mergeTwoLists(new ListNode, new ListNode);
	auto output2 = s.mergeTwoLists(new ListNode, new ListNode(0));

    cout << "*************output*************" << endl;
    printList(output);
    printList(output1);
    printList(output2);

    return 0;
}
